slo/rules.go

```diff
@@ -58,161 +58,110 @@ func (o Objective) Burnrates() (monitoringv1.RuleGroup, error) {
-				)),
-				For:    model.Duration(w.For).String(),
-				Labels: alertLabels,
-			}
-			rules = append(rules, r)
+		ruleGroup.Rules = append(ruleGroup.Rules, o.RecordingRules(matchers, groupingMap, burnrates, sloName)...)
+		if !o.Alerting.Disabled {
+			ruleGroup.Rules = append(ruleGroup.Rules, o.AlertingRules(matchers, groupingMap, ws, sloName)...)
 		}
 	}
 
 	if o.Indicator.Latency != nil && o.Indicator.Latency.Total.Name != "" {
 		matchers := o.Indicator.Latency.Total.LabelMatchers
-
 		groupingMap := map[string]struct{}{}
 		for _, g := range o.Indicator.Latency.Grouping {
 			groupingMap[g] = struct{}{}
 		}
+		ruleGroup.Rules = append(ruleGroup.Rules, o.RecordingRules(matchers, groupingMap, burnrates, sloName)...)
 
-		ruleLabels := o.commonRuleLabels(sloName)
-		for _, m := range matchers {
-			if m.Type == labels.MatchEqual && m.Name != labels.MetricName {
-				ruleLabels[m.Name] = m.Value
-			}
-		}
-		// Delete labels that are grouped as their value is part of the labels anyway
-		for g := range groupingMap {
-			delete(ruleLabels, g)
+		if !o.Alerting.Disabled {
+			ruleGroup.Rules = append(ruleGroup.Rules, o.AlertingRules(matchers, groupingMap, ws, sloName)...)
 		}
+	}
 
-		for _, br := range burnrates {
-			rules = append(rules, monitoringv1.Rule{
-				Record: o.BurnrateName(br),
-				Expr:   intstr.FromString(o.Burnrate(br)),
-				Labels: ruleLabels,
-			})
+	return ruleGroup, nil
+}
+
+func (o Objective) RecordingRules(matchers []*labels.Matcher, groupingMap map[string]struct{}, burnrates []time.Duration, sloName string) []monitoringv1.Rule {
+	rules := make([]monitoringv1.Rule, 0, len(burnrates))
+
+	ruleLabels := o.commonRuleLabels(sloName)
+	for _, m := range matchers {
+		if m.Type == labels.MatchEqual && m.Name != labels.MetricName {
+			ruleLabels[m.Name] = m.Value
 		}
+	}
+	// Delete labels that are grouped as their value is part of the labels anyway
+	for g := range groupingMap {
+		delete(ruleLabels, g)
+	}
 
-		var alertMatchers []string
-		for _, m := range matchers {
-			if m.Name == labels.MetricName {
-				continue
-			}
-			if _, ok := groupingMap[m.Name]; !ok {
-				if m.Type == labels.MatchRegexp || m.Type == labels.MatchNotRegexp {
-					continue
-				}
-			}
+	for _, br := range burnrates {
+		rules = append(rules, monitoringv1.Rule{
+			Record: o.BurnrateName(br),
+			Expr:   intstr.FromString(o.Burnrate(br)),
+			Labels: ruleLabels,
+		})
+	}
+
+	return rules
+}
 
-			alertMatchers = append(alertMatchers, m.String())
+func (o Objective) AlertingRules(matchers []*labels.Matcher, groupingMap map[string]struct{}, ws []Window, sloName string) []monitoringv1.Rule {
+	rules := make([]monitoringv1.Rule, 0, len(ws))
+	var alertMatchers []string
+	for _, m := range matchers {
+		if m.Name == labels.MetricName {
+			continue
 		}
-		alertMatchers = append(alertMatchers, fmt.Sprintf(`slo="%s"`, sloName))
-		sort.Strings(alertMatchers)
-		alertMatchersString := strings.Join(alertMatchers, ",")
```

Comment:
There is a slight difference between the rules for ratio and latency.
I believe these `alertMatchers` are the difference. 
It won't be possible to extract this logic into one function that works for both ratio and latency.
The regexp need to be handled differently in the latency SLO to allow for matching labels like `code=~"2.."`.

Reply:
Aaaaaah I see... That should be why all test that generates burn rates for latency are failing 🙁 

I copy pasted the logic from ratio and deleted from latency 🤦